### ISSUE TITLE
CSRF token in chsk-send!

### ISSUE THREAD

**Issue by coyotespike** [ptaoussanis/sente#267]
Hi! I've just started using `Sente`. Last year figuring out basic web comms was painful, and this looks great. I have a couple of questions. 

First, is there any situation when you _wouldn't_ want to use this library? Seems like it combines Ajax and WebSockets, so I can just use this for basically every client-server interaction?

I'm asking because Rich Hickey says that programmers know the benefits of everything and the tradeoffs of nothing. :-)

My second question is pretty specific. You can send a message in Sente: `(chsk-send! [:rente/testevent {:message "Hello socket Callback!"}]`. This has no CSRF token explicitly declared. I'm reading the docs and can't tell if  that's because it's sending the token back under the hood, or because it's not needed.

This is in contrast to this example from the docs:
```
           (sente/ajax-lite "/login"
             {:method :post
              :headers {:X-CSRF-Token (:csrf-token @chsk-state)}
              :params  {:user-id (str user-id)}}
```
 
Using the `ajax-lite` method, we do have to explicitly send the CSRF token, which is easily obtained from the `chsk-state` atom.

Also, if you're interested, I've written most of a basic tutorial, to extend the Readme.

**Comment by arichiardi** [comment]
Not an authoritative answer here, but the tradeoff I found and that you might encounter are a couple:

1) custom protocol for messages on the wire (a very simple one by the way, but you should be aware of it)
2) use of `core.async`, the server router runs in a `go` channel so you might want to avoid blocking in handlers and/or spawn new threads for potentially blocking computations.

It is not that big of a deal and everything should be documented (if not, it is coming, see mine and PR from others).

As for the `CSRF` token, websocket is still HTTP so you can/have to use the common ring approach to provide the token from the server to the client, there is a bit of description here: https://github.com/ptaoussanis/sente#security-csrf-protection (admittedly, it can be expanded a bit).

**Comment by ptaoussanis** [comment]
Hi there,

> First, is there any situation when you wouldn't want to use this library? Seems like it combines Ajax and WebSockets, so I can just use this for basically every client-server interaction?

That's an important question to ask.

Biggest thing that I've noticed is folks who switch to Sente tend to get accustomed to how quick/easy it is to build highly async applications. But that quickness/easiness is often deceptive, since there's more to building a solid async application than just the comms mechanism.

That is: the application and server design often need some careful thinking to figure out exactly how the application should behave in various edge cases (when connections drop, when broadcast messages aren't received because the user was driving through a tunnel, etc.).

Sente doesn't do anything to _solve_ questions like that (it can't, they'll be application specific), but because it does _help_ with many of these indirectly (e.g. by providing a certain degree of automatic retry, etc.) - so it can sometimes inadvertently encourage developers to put off the unpleasant task of actually carefully thinking through what the application design needs to be.

This sums up as:

* Highly async applications can be harder to build correctly than they first appear (in general, not a Sente issue)
* Sente can encourage a sense that everything will always just work since it provides partial but not total solutions to some common issues you'll run into when designing an async app

Another general concern (not to do with Sente in particular): highly async applications tend to put substantially more load on your servers. Again, a properly designed system will scale fine - but you need to think about your design.

Hope that helps!

> This has no CSRF token explicitly declared.

CSRF tokens are sent automatically when necessary, please see https://github.com/ptaoussanis/sente#security-csrf-protection for config info.

Note that CSRF tokens aren't necessary for individual WebSocket requests, only for the initial handshake. They are necessary for individual Ajax requests.

**Comment by coyotespike** [comment]
Hi arichiardi, thanks for chiming in. By custom protocol, do you mean EDN, or the usage of `event-msg-handler`, or something else?

The blocking issue is really good to know about. I'll look at your PR - maybe I can write a toy example to demonstrate for myself the issues to avoid.

For `CSRF`, I did read the description and examples but still don't quite get it. I really like that Sente insists on warning if CSRF isn't enabled. I have the Ring defaults on the back end, which include CSRF.

When using a channel socket to contact the server, though, I don't explicitly include the token. With `ajax-lite` we do. How does the `CSRF` token get to the server? 

Given that websocket is still HTTP, I guess the token must be getting passed somehow.

**Comment by coyotespike** [comment]
Hi @ptaoussanis, thanks for your thoughtful reply. That gives me a whole perspective on what to go read about and think about when designing an async application. 

And thanks for confirming the `CSRF` answer - I did read the source code but evidently didn't figure that part out. Perhaps [here](https://github.com/ptaoussanis/sente/blob/master/src/taoensso/sente.cljc#L870).

Between you and arichiardi, you've answered all my questions. :-)

**Comment by arichiardi** [comment]
See also Peter's more accurate answer but the only "difficulty" I found (which I documented in the PR) has to do with intergratiin tests and performance tests.
It has nothing to do with normal usage, the only thing to know is that sente does not send the payload you expect over the wire "as is". For example the content can be wrapped in an additional vector when the client uid is transported.
All this is useful if you want to emulate sente's behavior (which here at some point I wanted to do, if only for see what actually happens under the hood). In hindsight, it is not even a tradeoff, but better to be aware of it then not, don't you agree? :smile:
